include/pm/device.h

```diff
@@ -151,6 +151,22 @@ struct pm_device {
  */
 #define Z_PM_DEVICE_NAME(dev_name) _CONCAT(__pm_device__, dev_name)
 
+/**
+ * @brief Define device PM slot.
+ *
+ * This macro defines a pointer to a device in the z_pm_device_slots region.
+ * When invoked for each device with PM, it will effectively result in a device
+ * pointer array with the same size of the actual devices with PM enabled. This
+ * is used internally by the PM subsystem to keep track of suspended devices
+ * during system power transitions.
+ *
+ * @param dev_name Device name.
+ */
+#define Z_PM_DEVICE_DEFINE_SLOT(dev_name)				\
+	static const Z_DECL_ALIGN(struct device *)			\
+	_CONCAT(Z_PM_DEVICE_NAME(dev_name), slot) __used		\
+	__attribute__((__section__(".z_pm_device_slots")))
```

Comment:
Do you also need to change `pm_suspend_devices()`? That routine loops over all devices, regardless of whether the device enables PM or not.

I think it makes sense to split this commit into a separate PR as this is doing more than name changes like the other commits.

Reply:
Even though it iterates over all devices, only those with `pm != NULL` have a chance to use a slot, since others are ignored. I'll drop this commit and open a new PR, since I have more stuff to push in that area.

Reply:
https://github.com/zephyrproject-rtos/zephyr/pull/40576